Automattic/woocommerce-payments
# HPOS: Fix subscriptions remaining on-hold after WCPay Subscription renewal

Fixes #5264 

#### Changes proposed in this Pull Request

On HPOS environments, renewing WCPay Subscriptions would result in the subscription remaining on-hold after the renewal order was successfully paid. 

After some debugging I discovered that this error occurs because WC on HPOS always saves the status, even it hasn't changed. WC Core issue: https://github.com/woocommerce/woocommerce/issues/35951

In this context, the subscription was on-hold because of the following flow:

```php
function handle_invoice_paid()
│ $subscription_a = ...
│ $subscription_a->update_status( 'on-hold' );
│
○ $renewal_order->payment_complete();
│╲
│ ╲
│  ╲
│   ○ function WC_Subscriptions_Renewal_Order::maybe_record_subscription_payment()
│   │ 
│   $subscription_b = ...
│   $subscription_b->update_status( 'active' ) 
│
|
○ function mark_pending_invoice_paid_for_subscription()
|
$subscription_a->save() // overrides the 'active' status with 'on-hold' because that's the status $subscription_a had.
```

To summerise, a separate flow changes the status from on-hold to active, then the original instance that still thinks the subscription is on-hold is saved and that overrides that active status.

This PR fixes that by getting a fresh instance of the subscription after calling `$order->payment_complete()` so that the original instance now knows it has an active status.

#### Testing instructions





Legit testing:

* Enable HPOS
* Create a daily subscription
* Purchase the daily subscription using WCPay
* Wait until the subscription renews tomorrow.
* Confirm that the subscription after renewing has an active status.
    - On trunk the subscription stays on-hold.

Hacky way to test this:

* Enable HPOS
* Create a daily subscription
* Purchase the daily subscription using WCPay
* Run the following code snippet.
     - On `develop` the subscription will remain on-hold
     - On this branch the subscription will renew and have an active status. 

```php
$subscription =  wcs_get_subscription( 1006 );

$mocked_webhook_data = [
    'data'       => [
        'object' => [
            'subscription' => WC_Payments_Subscription_Service::get_wcpay_subscription_id( $subscription ),
            'id'           => 'in_' . gmdate( "U" ), // We need a unique invoice ID to create a new renewal order
        ],
    ],
];

WC_Payments_Subscriptions::get_event_handler()->handle_invoice_paid( $mocked_webhook_data );
```

-------------------


**Post merge**
